src/platform_impl/android/mod.rs

```diff
@@ -333,36 +335,80 @@ impl EventLoop {
                     _ => None,
                 };
 
-                if let Some(pointers) = pointers {
-                    for pointer in pointers {
-                        let tool_type = pointer.tool_type();
-                        let position =
-                            PhysicalPosition { x: pointer.x() as _, y: pointer.y() as _ };
-                        trace!(
-                            "Input event {device_id:?}, {action:?}, loc={position:?}, \
-                             pointer={pointer:?}, tool_type={tool_type:?}"
-                        );
-                        let finger_id = event::FingerId(FingerId(pointer.pointer_id()));
-                        let force = Some(Force::Normalized(pointer.pressure() as f64));
-
-                        match action {
-                            MotionAction::Down | MotionAction::PointerDown => {
-                                let event = event::WindowEvent::PointerEntered {
-                                    device_id,
-                                    position,
-                                    kind: match tool_type {
-                                        android_activity::input::ToolType::Finger => {
-                                            event::PointerKind::Touch(finger_id)
-                                        },
-                                        // TODO mouse events
-                                        android_activity::input::ToolType::Mouse => continue,
-                                        _ => event::PointerKind::Unknown,
+                for pointer in pointers.into_iter().flatten() {
+                    let tool_type = pointer.tool_type();
+                    let position = PhysicalPosition { x: pointer.x() as _, y: pointer.y() as _ };
+                    trace!(
+                        "Input event {device_id:?}, {action:?}, loc={position:?}, \
+                         pointer={pointer:?}, tool_type={tool_type:?}"
+                    );
+                    let finger_id = event::FingerId(FingerId(pointer.pointer_id()));
+                    let force = Some(Force::Normalized(pointer.pressure() as f64));
+
+                    match action {
+                        MotionAction::Down | MotionAction::PointerDown => {
+                            let primary = action == MotionAction::Down;
```

Comment:
```suggestion
                            let primary = matches!(action, MotionAction::Cancel | MotionAction::Down);
```
I *think* this should account for `ACTION_CANCEL` as well?
The [documentation](https://developer.android.com/reference/android/view/MotionEvent#ACTION_CANCEL) says the following:
> You should treat this as an up event, but not perform any action that you normally would.

So I'm assuming you won't get a `ACTION_UP` anymore?

Reply:
Don't mind checking it as well just to be safe.

Reply:
Guess this suggestion should have been on `MotionAction::Up`, at least that's where Kirill applied it (and the match arm already included it). Would love to see for how many fingers `ACTION_CANCEL` is raised, probably each of them that is down, and requiring us to keep track of `self.primary_pointer` just like for the `Move` action?

Reply:
> [..], and requiring us to keep track of `self.primary_pointer` just like for the `Move` action?

Kirill seems to have thought of that as well, so I hope we really covered everything now.